src/gluonts/torch/model/deepar/deepar_network.py

```diff
@@ -0,0 +1,430 @@
+        embedding_dimension: Optional[List[int]] = None,
+        num_layers: int = 2,
+        num_cells: int = 40,
+        cell_type: str = "LSTM",
+        context_length: Optional[int] = None,
+        distr_output: DistributionOutput = StudentTOutput(),
+        dropout_rate: float = 0.1,
+        lags_seq: Optional[List[int]] = None,
+        scaling: bool = True,
+        num_parallel_samples: int = 100,
+    ) -> None:
+        super().__init__()
+        self.num_layers = num_layers
+        self.num_cells = num_cells
+        self.cell_type = cell_type
+        self.context_length = context_length or prediction_length
+        self.prediction_length = prediction_length
+        self.dropout_rate = dropout_rate
+        self.num_feat_dynamic_real = num_feat_dynamic_real
+        self.num_feat_static_cat = num_feat_static_cat
+        self.num_feat_static_real = num_feat_static_real
+        self.cardinality = cardinality
+        self.embedding_dimension = (
+            embedding_dimension
+            if embedding_dimension is not None or self.cardinality is None
+            else [min(50, (cat + 1) // 2) for cat in self.cardinality]
+        )
+        self.scaling = scaling
+        self.lags_seq = lags_seq or get_lags_for_frequency(
+            freq_str=freq, lag_ub=self.context_length
+        )
+        self.num_parallel_samples = num_parallel_samples
+        self.history_length = self.context_length + max(self.lags_seq)
+        # for decoding the lags are shifted by one, at the first time-step
+        # of the decoder a lag of one corresponds to the last target value
+        self.shifted_lags = [l - 1 for l in self.lags_seq]
+        self.distr_output = distr_output
+        rnn = {"LSTM": nn.LSTM, "GRU": nn.GRU}[self.cell_type]
+        self.rnn = rnn(
+            input_size=self._rnn_input_size,
+            hidden_size=num_cells,
+            num_layers=num_layers,
+            dropout=dropout_rate,
+            batch_first=True,
+        )
+        self.target_shape = distr_output.event_shape
+        self.proj_distr_args = distr_output.get_args_proj(num_cells)
+        if self.cardinality is not None:
+            self.embedder = FeatureEmbedder(
+                cardinalities=self.cardinality,
+                embedding_dims=self.embedding_dimension,
+            )
+        else:
+            self.embedder = None
+        if scaling:
+            self.scaler = MeanScaler(keepdim=True)
+        else:
+            self.scaler = NOPScaler(keepdim=True)
+
+    @property
+    def _rnn_input_size(self) -> int:
+        return (
+            1  # TODO adjust this for multivariate target?
+            + len(self.lags_seq)
+            + sum(self.embedding_dimension)
+            + self.num_feat_dynamic_real
+            + self.num_feat_static_real
+        )
+
+    @property
+    def _past_length(self) -> int:
+        return self.context_length + max(self.lags_seq)
+
+    @staticmethod
+    def get_lagged_subsequences(
```

Comment:
Not for this PR, but maybe we can pull this out and use the `@singledispatch` mechanism for having this in torch/mxnet?